server/questions.js

```diff
@@ -254,4 +255,252 @@ exports.verifyAnswer = function(question, answer) {
+                    voteValue = 0;
+                    update.$pull = { 'comments.$.dislikes': userId };
+                    update.$inc = { 'comments.$.dislikesCount': -1 };
+
+                    if (vote === 1) {
+						updatedLikesCount++;
+                        voteValue = 1;
+                        update.$push = { 'comments.$.likes': userId };
+                        update.$inc['comments.$.likesCount'] = 1;
+                    }
+                }
+
+                if (!userUpVoted && !userDownVoted) {
+                    if (vote === 1) {
+						updatedLikesCount++;
+                        update.$push = { 'comments.$.likes': userId };
+                        update.$inc = { 'comments.$.likesCount': 1 };
+                    }
+
+                    if (vote === -1) {
+						updatedDisLikesCount++;
+                        update.$push = { 'comments.$.dislikes': userId };
+                        update.$inc = { 'comments.$.dislikesCount': 1 };
+                    }
+
+                    voteValue = vote;
+                }
+
+                return db.updateQuestionByQuery(query, update, function (err, result) {
+                    return callback(err, {
+						result: result,
+						voteValue: voteValue,
+						likesCount: updatedLikesCount,
+						dislikesCount: updatedDisLikesCount
+					});
+                });
+            }
+        }
+
+        return callback('Could not submit the vote on the comment', null);
+    });
+}
+
+// vote on a reply
+exports.voteReply = function (replyId, vote, userId, callback) {
+    var query = {'comments.replies._id': replyId};
+    var update = {};
+    var voteValue = -2;
+
+    db.lookupQuestion(query, function(err, question){
+        if (err) {
+            return callback(err, null);
+        }
+
+        if (!question) {
+            return callback('Question not found based on replyId', null);
+        }
+
+        // TODO: optimize this using mongodb projections
+        var comments = question.comments;
+        for (var j in comments) {
+            var replies = comments[j].replies;
+            for (var i in replies) {
+                if (replies[i]._id === replyId) {
+                    var userUpVoted = replies[i].likes.indexOf(userId) !== -1;
+                    var userDownVoted = replies[i].dislikes.indexOf(userId) !== -1;
+                    var updatedLikesCount = replies[i].likesCount;
+                    var updatedDisLikesCount = replies[i].dislikesCount;
+
+                    if (userUpVoted && userDownVoted) {
+                        return callback('User liked and disliked a comment at the sametime', null);
+                    }
+
+                    if (userUpVoted) {
+                        updatedLikesCount--;
+                        voteValue = 0;
+                        var commentsreplieslikes = 'comments.'+j+'.replies.'+i+'.likes';
+                        var commentsreplieslikesCount = 'comments.'+j+'.replies.'+i+'.likesCount';
```

Comment:
I do not understand what is going on here with j and i. Add comments and change j and i to better variable names